include/inviwo/core/datastructures/geometry/decoratedmesh.h

```diff
@@ -0,0 +1,432 @@
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#ifndef IVW_DECORATEDMESH_H
+#define IVW_DECORATEDMESH_H
+
+#include <inviwo/core/common/inviwocoredefine.h>
+#include <inviwo/core/common/inviwo.h>
+
+#include <inviwo/core/datastructures/geometry/geometrytype.h>
+#include <inviwo/core/datastructures/geometry/mesh.h>
+#include <inviwo/core/datastructures/buffer/buffer.h>
+#include <inviwo/core/datastructures/buffer/bufferram.h>
+#include <inviwo/core/datastructures/buffer/bufferramprecision.h>
+
+namespace inviwo {
+
+/**
+ * \defgroup decoratedmesh DecoratedMesh
+ * \ingroup datastructures
+ *
+ * \copydetails DecoratedMesh
+ *
+ *
+ */
+
+/**
+ * Namespace for buffer traits
+ * \ingroup decoratedmesh
+ */
+namespace buffertraits {
+
+namespace detail {
+template <typename T, typename... Ts>
+struct get_index;
+
+template <typename T, typename... Ts>
+struct get_index<T, T, Ts...> : std::integral_constant<std::size_t, 0> {};
+
+template <typename T, typename Tail, typename... Ts>
+struct get_index<T, Tail, Ts...>
+    : std::integral_constant<std::size_t, 1 + get_index<T, Ts...>::value> {};
+
+}  // namespace detail
+
+/**
+ * \ingroup decoratedmesh
+ * Struct used to tell DecoratedMesh what buffers to create.
+ * The first two template parameters, 'typename T, unsigned DIM' is used rather than
+ * ���Vector<DIM,T>��� or glm::TvecX due to compiler issues related to the clone function.
+ * Similarly, The template parameter ���int attrib��� needs can't be a strongly typed enum
+ * (inviwo::BufferType) and hence it is an int.
+ *
+ * \see DecoratedMesh
+ * \see decoratedmesh
+ */
+template <typename T, unsigned DIM, int attrib, int location = attrib>
+struct BufferTrait {
+    using type = Vector<DIM, T>;
+    static inviwo::Mesh::BufferInfo bi() {
+        return {static_cast<inviwo::BufferType>(attrib), location};
+    }
+
+#if _MSC_VER <= 1900
```

Comment:
What is it that does not work in vs2015

Reply:
Should document this. But: 

I could not get the code for addVertex([custom parameters]) to compile. The following code in various variation refused to compile. I do not remember the exact error messages. 
```
template <typename T>
using TypeAlias = typename T::type;
...
void addVertex(TypeAlias<BufferTraits>... args) { addVertexImpl<0>(args...); }
```